hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/InPredicateIntegrationTest.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.sql.HazelcastSqlException;
+import com.hazelcast.sql.SqlColumnType;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.impl.expression.ExpressionTestSupport;
+import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Assert;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runners.Parameterized;
+
+import java.time.LocalDate;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class InPredicateIntegrationTest extends ExpressionTestSupport {
+    protected String longList = "(0, 1, 2, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 25, 28, 31, 35)";
+
+    @Test
+    public void inPredicateWithDifferentListLengthTest() {
+        putAll(0, 1, 25, 30);
+        checkValues(sqlQuery("IN (0, 1, 2)"), SqlColumnType.INTEGER, new Integer[]{0, 1});
+        checkValues(sqlQuery("IN (?, ?, ?)"), SqlColumnType.INTEGER, new Integer[]{0, 1}, 0, 1, 2);
+        checkValues(sqlQuery("IN " + longList), SqlColumnType.INTEGER, new Integer[]{25, 0, 1});
+    }
+
+    @Test
+    public void notInPredicateWithDifferentListLengthTest() {
+        putAll(0, 1, 3);
+        checkValues(sqlQuery("NOT IN (0, 1, 2)"), SqlColumnType.INTEGER, new Integer[]{3});
+        checkValues(sqlQuery("NOT IN (?, ?, ?)"), SqlColumnType.INTEGER, new Integer[]{3}, 0, 1, 2);
+        checkValues(sqlQuery("NOT IN " + longList), SqlColumnType.INTEGER, new Integer[]{3});
+    }
+
+    @Test
+    public void inPredicateWithRawNullTest() {
+        putAll(0, 1, 2);
+        checkValues(sqlQuery("IN (NULL, 0, 1)"), SqlColumnType.INTEGER, new Integer[]{0, 1});
+        checkValues(sqlQuery("IN (0, NULL, NULL, NULL, 2)"), SqlColumnType.INTEGER, new Integer[]{2, 0});
+        checkValues(sqlQuery("NOT IN (0, NULL, NULL, NULL, 2)"), SqlColumnType.INTEGER, new Integer[]{1});
+        checkValues(sqlQuery("NOT IN (NULL, 2, 1)"), SqlColumnType.INTEGER, new Integer[]{0});
```

Comment:
The following test fails here:
```
checkValues("SELECT this FROM map WHERE true <> (this IN (NULL, 2, 1))", SqlColumnType.INTEGER, new Integer[]{0});
```
The result should be empty, but it's `[0]`. It's the consequence of removing the `null` item from the value list. When we remove it, for `this==0`, the `IN` expression returns `false` instead of `unknown`. And `true <> false` is `true`, while `true <> unknown` is `unknown`.

![image](https://user-images.githubusercontent.com/9109877/112664946-d06df500-8e5a-11eb-8f41-d82593f89694.png)

Reply:
This still needs addressing...